test/test_compile.py

```diff
@@ -774,26 +774,65 @@ def call(x, td):
 
 
 @pytest.mark.skipif(not _v2_5, reason="Requires PT>=2.5")
+@pytest.mark.parametrize("strict", [True, False])
 class TestExport:
-    def test_export_module(self):
+    def test_export_module(self, strict):
         torch._dynamo.reset_code_caches()
         tdm = Mod(lambda x, y: x * y, in_keys=["x", "y"], out_keys=["z"])
         x = torch.randn(3)
         y = torch.randn(3)
-        out = torch.export.export(tdm, args=(), kwargs={"x": x, "y": y})
+        out = torch.export.export(tdm, args=(), kwargs={"x": x, "y": y}, strict=strict)
         assert (out.module()(x=x, y=y) == tdm(x=x, y=y)).all()
 
-    def test_export_seq(self):
+    def test_export_seq(self, strict):
         torch._dynamo.reset_code_caches()
         tdm = Seq(
             Mod(lambda x, y: x * y, in_keys=["x", "y"], out_keys=["z"]),
             Mod(lambda z, x: z + x, in_keys=["z", "x"], out_keys=["out"]),
         )
         x = torch.randn(3)
         y = torch.randn(3)
-        out = torch.export.export(tdm, args=(), kwargs={"x": x, "y": y})
+        out = torch.export.export(tdm, args=(), kwargs={"x": x, "y": y}, strict=strict)
         torch.testing.assert_close(out.module()(x=x, y=y), tdm(x=x, y=y))
 
+    @pytest.mark.parametrize(
+        "same_shape,dymanic_shape", [[True, True], [True, False], [False, True]]
+    )
+    def test_td_output(self, strict, same_shape, dymanic_shape):
+        # This will only work when the tensordict is pytree-able
+        class Test(torch.nn.Module):
+            def forward(self, x: torch.Tensor, y: torch.Tensor):
+                return TensorDict(
+                    {
+                        "x": x,
+                        "y": y,
+                    },
+                    batch_size=x.shape[0],
+                )
+
+        test = Test()
+        if same_shape:
+            x, y = torch.zeros(5, 100), torch.zeros(5, 100)
+        else:
+            x, y = torch.zeros(2, 100), torch.zeros(2, 100)
+        if dymanic_shape:
+            kwargs = {
+                "dynamic_shapes": {
+                    "x": {0: torch.export.Dim("batch"), 1: torch.export.Dim("time")},
+                    "y": {0: torch.export.Dim("batch"), 1: torch.export.Dim("time")},
+                }
+            }
+        else:
+            kwargs = {}
+
+        result = torch.export.export(test, args=(x, y), strict=False, **kwargs)
+        export_mod = result.module()
+        x_new, y_new = torch.zeros(5, 100), torch.zeros(5, 100)
+        export_test = export_mod(x_new, y_new)
+        eager_test = test(x_new, y_new)
+        assert eager_test.batch_size == export_test.batch_size
```

Comment:
@ezyang this seems to be lost when using dynamic shape.
The batch size `[s0]` becomes `[]` when using dynamic shapes and when the 2nd output shape mismatches the 1st. 
We do get a warning though
```
W0920 10:19:28.564000 20340 torch/fx/experimental/symbolic_shapes.py:5136] Ignored guard Eq(s0, 5) == False, this could result in accuracy problems
```

Reply:
So, there's something a bit nontrivial going on here. In torch.compile eager, if we produce a fresh TensorDict and that TensorDict holds a list of dynamic ints, then in the residual bytecode we have to construct the TensorDict and also put in the freshly computed dynamic shapes from the FX graph (that has some int outputs now). So actually building a TensorDict isn't just a matter of putting in the right tensors, you also have to put some ints in too. Does this work?

Assuming this does work, export also has to be setup to do the same thing as well. It wouldn't be surprising if it didn't. In particular, if all export is doing is a pytree unflatten on Tensor leaves, the batch size won't be modified at all. To address this, we need to fix the export bug. But I also saw the comment about TensorDict not being pytree-able, so I am uncertain about the status there.

If you want to workaround, perhaps batch size can store rank instead of size and lazily compute it from tensor if it's not set? Better to fix things though. Just not sure what you expect to work and not work.

Reply:
> Assuming this does work, export also has to be setup to do the same thing as well. It wouldn't be surprising if it didn't. In particular, if all export is doing is a pytree unflatten on Tensor leaves, the batch size won't be modified at all. To address this, we need to fix the export bug. But I also saw the comment about TensorDict not being pytree-able, so I am uncertain about the status there.



TensorDict is pytreeable but you can deactivate it, this is what the comment is about (don't do it or the test will fail)

Reply:
Here's what works and what doesn't
```python
        class Test(torch.nn.Module):
            def forward(self, x: torch.Tensor, y: torch.Tensor):
                return TensorDict(
                    {
                        "x": x,
                        "y": y,
                    },
                    batch_size=x.shape[0],
                )
     x, y = torch.zeros(5, 100), torch.zeros(5, 100)
     result = torch.export.export(test, args=(x, y), strict=False, dynamic_shapes={
                    "x": {0: torch.export.Dim("batch"), 1: torch.export.Dim("time")},
                    "y": {0: torch.export.Dim("batch"), 1: torch.export.Dim("time")},
                })
    result = torch.export.export(test, args=(x, y), strict=False, **kwargs)
    export_mod = result.module()
    x_new, y_new = torch.zeros(5, 100), torch.zeros(5, 100)
    export_test = export_mod(x_new, y_new)
    eager_test = test(x_new, y_new)
    assert torch.Size([5]) == eager_test.batch_size == export_test.batch_size # Works because x and x_new have the same shape

    x_new, y_new = torch.zeros(2, 100), torch.zeros(2, 100)
    export_test = export_mod(x_new, y_new)
    eager_test = test(x_new, y_new)
    assert torch.Size([2]) == eager_test.batch_size == export_test.batch_size # Fails! now export_test.batch_size is torch.Size([])
```
So it's a weird behaviour, the SymInt just vanished into thin air in the second case